src/analyses/cfg_dominators.h

```diff
@@ -44,180 +187,276 @@ class cfg_dominators_templatet
         }
-        else if(*o_it<*n_it)
-          ++o_it;
-        else
-        {
-          ++n_it;
-          ++o_it;
+        ancestor[v] = ancestor[ancestor[v]];
+      }
+    }
+
+    node_indext eval(node_indext v) {
+      if (ancestor[v] == 0) {
+        return label[v];
+      }
+      compress(v);
+      if (semi[label[ancestor[v]]] >= semi[label[v]]) {
+        return label[v];
+      }
+      return label[ancestor[v]];
+    }
+
+    void link(node_indext v, node_indext w) {
+      node_indext s = w;
+      while (semi[label[w]] < semi[label[child[s]]]) {
+        if (size[s] + size[child[child[s]]] >= 2 * size[child[s]]) {
+          ancestor[child[s]] = s;
+          child[s] = child[child[s]];
+        } else {
+          size[child[s]] = size[s];
+          s = ancestor[s] = child[s];
         }
       }
+      label[s] = label[w];
+      size[v] = size[v] + size[w];
+      if (size[v] < 2 * size[w]) {
+        std::swap(s, child[v]);
+      }
+      while (s != 0) {
+        ancestor[s] = v;
+        s = child[s];
+      }
+    }
 
-      while(n_it!=node.dominators.end())
-      {
-        if(*n_it==current)
-          ++n_it;
-        else
-        {
-          changed=true;
-          node.dominators.erase(n_it++);
+    void assign_dominators(node_indext root) {
+      std::stack<node_indext> work;
+      work.push(root);
+      while (!work.empty()) {
+        node_indext v = work.top();
+        work.pop();
+        if (cfg_dominators.cfg[v - 1].dominators.empty()) {
+          cfg_dominators.cfg[v - 1].dominators =
+              target_sett(cfg_dominators.dominators_data, v - 1);
+          for_each_successor(v, [&](node_indext w) { work.push(w); });
         }
       }
     }
 
-    if(changed) // fixed point for node reached?
-    {
-      for(const auto &edge : (post_dom ? node.in : node.out))
-      {
-        worklist.push_back(cfg[edge.first].PC);
+    bool is_unreachable(node_indext index) {
+      return semi[index] == 0;
+    }
+
+    template<typename Action>
+    void for_each_successor(node_indext node_index, Action action) {
+      // the -1 / +1 adjusts indices from 1 based to 0 based and back
+      auto ix = node_index - 1;
+      for (auto const &next :
+          post_dom ? cfg_dominators.cfg.in(ix) : cfg_dominators.cfg.out(ix)) {
+        action(next.first + 1);
+      }
+    }
+
+    template<typename Action>
+    void for_each_predecessor(node_indext node_index, Action action) {
+      auto ix = node_index - 1;
+      for (auto const &prev :
+          post_dom ? cfg_dominators.cfg.out(ix) : cfg_dominators.cfg.in(ix)) {
+        action(prev.first + 1);
+      }
+    }
+  };
+};
+
+template<class P, class T, bool post_dom>
+void cfg_dominators_templatet<P, T, post_dom>::fixedpointt::fixedpoint(
+    P &program) {
+  // Dominator Tree according to Lengauer and Tarjan
+  // "A fast algorithm for finding dominators in a flow graph"
+  // This is ununderstandable without reading the paper!
```

Comment:
That's not a good job done, then.

What should happen: surely it is possible to give the gist of the algorithm in a few sentences. Each of the key steps of the algorithm should then be commented on below.